core/scheduler/service/replay_worker.go

```diff
@@ -28,281 +36,229 @@ type ReplayScheduler interface {
-	updatedRuns := scheduler.JobRunStatusList(replayReq.Runs).MergeWithUpdatedRuns(updatedReplayMap)
+		// pick runs to be triggered
+		statesForReplay := []scheduler.State{scheduler.StatePending, scheduler.StateMissing}
+		toBeReplayedRuns := scheduler.JobRunStatusList(syncedRunStatus).GetSortedRunsByStates(statesForReplay)
+		if len(toBeReplayedRuns) == 0 {
+			continue
+		}
 
-	replayedRuns := scheduler.JobRunStatusList(updatedRuns).GetSortedRunsByStates([]scheduler.State{scheduler.StateInProgress})
-	toBeReplayedRuns := scheduler.JobRunStatusList(updatedRuns).GetSortedRunsByStates([]scheduler.State{scheduler.StatePending})
+		// execute replay run on scheduler
+		var updatedRuns []*scheduler.JobRunStatus
+		if replayWithRun.Replay.Config().Parallel {
+			if err := w.replayRunOnScheduler(ctx, jobCron, replayWithRun.Replay, toBeReplayedRuns...); err != nil {
+				return err
+			}
+			updatedRuns = scheduler.JobRunStatusList(toBeReplayedRuns).OverrideWithStatus(scheduler.StateInProgress)
+		} else { // sequential should work when there's no in_progress state on existing runs
+			inProgressRuns := scheduler.JobRunStatusList(syncedRunStatus).GetSortedRunsByStates([]scheduler.State{scheduler.StateInProgress})
+			if len(inProgressRuns) > 0 {
+				w.logger.Info("[ReplayID: %s] %d run is in progress, skip sequential iteration", replayWithRun.Replay.ID(), len(inProgressRuns))
+				continue
+			}
+			if err := w.replayRunOnScheduler(ctx, jobCron, replayWithRun.Replay, toBeReplayedRuns[0]); err != nil {
+				return err
+			}
+			updatedRuns = scheduler.JobRunStatusList(toBeReplayedRuns[:1]).OverrideWithStatus(scheduler.StateInProgress)
+		}
 
-	replayState := scheduler.ReplayStatePartialReplayed
-	if len(replayedRuns) == 0 && len(toBeReplayedRuns) > 0 {
-		runToReplay := toBeReplayedRuns[0]
-		if err := w.replayRunOnScheduler(ctx, replayReq, jobCron, runToReplay); err != nil {
+		// update runs status
+		if err := w.replayRepo.UpdateReplay(ctx, replayWithRun.Replay.ID(), scheduler.ReplayStateInProgress, updatedRuns, ""); err != nil {
+			w.logger.Error("unable to update replay runs for replay_id [%s]: %s", replayWithRun.Replay.ID(), err)
 			return err
 		}
-
-		updatedReplayMap[runToReplay.ScheduledAt] = scheduler.StateInProgress
-		updatedRuns = scheduler.JobRunStatusList(updatedRuns).MergeWithUpdatedRuns(updatedReplayMap)
 	}
+}
 
-	pendingRuns := scheduler.JobRunStatusList(updatedRuns).GetSortedRunsByStates([]scheduler.State{scheduler.StatePending})
-	if len(pendingRuns) == 0 {
-		replayState = scheduler.ReplayStateReplayed
+func (w *ReplayWorker) finishReplay(ctx context.Context, replayID uuid.UUID, syncedRunStatus []*scheduler.JobRunStatus) error {
+	replayState := scheduler.ReplayStateSuccess
+	msg := ""
+	if isAnyFailure(syncedRunStatus) {
+		replayState = scheduler.ReplayStateFailed
+		msg = "replay is failed due to some of runs are in failed state" // TODO: find out how to pass the meaningful failed message here
```

Comment:
should we update it?

Reply:
updated this to includes the summary of runs status